azkaban-web-server/src/main/java/azkaban/flowtrigger/TriggerInstanceProcessor.java

```diff
@@ -0,0 +1,116 @@
+
+  private static final Logger logger = LoggerFactory.getLogger(TriggerInstanceProcessor.class);
+  private static final String FAILURE_EMAIL_SUBJECT = "flow trigger for %s "
+      + "cancelled from %s";
+  private static final String FAILURE_EMAIL_BODY = "Your flow trigger cancelled [id: %s]";
+
+  private final ExecutorManager executorManager;
+  private final FlowTriggerInstanceLoader flowTriggerInstanceLoader;
+  private final Emailer emailer;
+
+  @Inject
+  public TriggerInstanceProcessor(final ExecutorManager executorManager,
+      final FlowTriggerInstanceLoader flowTriggerInstanceLoader,
+      final Emailer emailer) {
+    Preconditions.checkNotNull(executorManager);
+    Preconditions.checkNotNull(flowTriggerInstanceLoader);
+    Preconditions.checkNotNull(emailer);
+    this.emailer = emailer;
+    this.executorManager = executorManager;
+    this.flowTriggerInstanceLoader = flowTriggerInstanceLoader;
+  }
+
+  private void executeFlowAndUpdateExecID(final TriggerInstance triggerInst) {
+    try {
+      final Project project = triggerInst.getProject();
+      final Flow flow = FlowUtils.getFlow(project, triggerInst.getFlowId());
+      final ExecutableFlow executableFlow = FlowUtils.createExecutableFlow(project, flow);
+      // execute the flow with default execution option(concurrency option being "ignore
+      // currently running")
+      this.executorManager.submitExecutableFlow(executableFlow, triggerInst.getSubmitUser());
+      triggerInst.setFlowExecId(executableFlow.getExecutionId());
+      this.flowTriggerInstanceLoader.updateAssociatedFlowExecId(triggerInst);
+    } catch (final Exception ex) {
+      logger.error("exception when executing the associate flow and updating flow exec id", ex);
+      //todo chengren311: should we swallow the exception or notify user
+    }
+  }
+
+  private String generateFailureEmailSubject(final TriggerInstance triggerInstance) {
+    final String flowFullName =
+        triggerInstance.getProjectName() + "." + triggerInstance.getFlowId();
+    return String.format(FAILURE_EMAIL_SUBJECT, flowFullName, this.emailer.getAzkabanName());
+  }
+
+  private String generateFailureEmailBody(final TriggerInstance triggerInstance) {
+    final String triggerInstFullName =
+        triggerInstance.getProjectName() + "." + triggerInstance.getFlowId();
+    return String.format(FAILURE_EMAIL_BODY, triggerInstFullName);
+  }
+
+  private void sendFailureEmailIfConfigured(final TriggerInstance triggerInstance) {
+    final List<String> failureEmails = triggerInstance.getFailureEmails();
+    if (!failureEmails.isEmpty()) {
+      this.emailer.sendEmail(failureEmails, generateFailureEmailSubject(triggerInstance),
+          generateFailureEmailBody(triggerInstance));
+    }
+  }
+
+  /**
+   * Process the case where status of trigger instance becomes success
+   */
+  public void processSucceed(final TriggerInstance triggerInst) {
+    logger.debug("process succeed for " + triggerInst);
```

Comment:
This just reminds me that did we consider publishing events to Azkaban Project Events page?

Reply:
that's good idea i will add it to todo.